lib/src/mt_ptp.c

```diff
@@ -42,6 +53,141 @@ static inline void ptp_timesync_lock(struct mt_ptp_impl* ptp) { /* todo */
 static inline void ptp_timesync_unlock(struct mt_ptp_impl* ptp) { /* todo */
 }
 
+static inline double pi_sample(struct mt_pi_servo* s, double offset, double local_ts,
+                               enum servo_state* state) {
+  double ppb = 0.0;
+
+  switch (s->count) {
+    case 0:
+      s->offset[0] = offset;
+      s->local[0] = local_ts;
+      *state = UNLOCKED;
+      s->count = 1;
+      break;
+    case 1:
+      s->offset[1] = offset;
+      s->local[1] = local_ts;
+      *state = UNLOCKED;
+      s->count = 2;
+      break;
+    case 2:
+      s->drift += (s->offset[1] - s->offset[0]) / (s->local[1] - s->local[0]);
+      *state = UNLOCKED;
+      s->count = 3;
+      break;
+    case 3:
+      *state = JUMP;
+      s->count = 4;
+      break;
+    case 4:
+      s->drift += KI * offset;
+      ppb = KP * offset + s->drift;
+      *state = LOCKED;
+      break;
+  }
+
+  return ppb;
+}
+
+void ptp_adj_system_clock_time(struct mt_ptp_impl* ptp, int64_t delta) {
+  struct timex adjtime;
+  int sign = 1;
+
+  if (delta < 0) {
+    sign = -1;
+    delta *= -1;
+  }
+
+  memset(&adjtime, 0, sizeof(adjtime));
+  adjtime.modes = ADJ_SETOFFSET | ADJ_NANO;
+  adjtime.time.tv_sec = sign * (delta / NS_PER_S);
+  adjtime.time.tv_usec = sign * (delta % NS_PER_S);
+  if (adjtime.time.tv_usec < 0) {
+    adjtime.time.tv_sec -= 1;
+    adjtime.time.tv_usec += 1000000000;
+  }
+
+  clock_adjtime(CLOCK_REALTIME, &adjtime);
```

Comment:
Does this operation need a root access?

Reply:
> Does this operation need a root access?

Yes, the clock_adjtime need a root access. I think the ptp adjustment function rte_eth_timesync_adjust_time also need a root access. The phc2sys start with ptp enable. So, it doesn't require additional root access.

Reply:
rte_eth_timesync_adjust_time is pure dpdk user driver settings, it can run from non-root user.

Reply:
> rte_eth_timesync_adjust_time is pure dpdk user driver settings, it can run from non-root user.

Add a command line argument "--phc2sys" to enable phc2sys. So, user can enable/disable the feature based on their requirement.